mempool/src/tests/multi_node_test.rs

```diff
@@ -0,0 +1,941 @@
+    channel::mpsc::{self, unbounded, UnboundedReceiver},
+    executor::block_on,
+    future::FutureExt,
+    StreamExt,
+};
+use netcore::transport::ConnectionOrigin;
+use network::{
+    peer_manager::{
+        conn_notifs_channel, ConnectionNotification, ConnectionRequestSender,
+        PeerManagerNotification, PeerManagerRequest, PeerManagerRequestSender,
+    },
+    protocols::network::{NetworkEvents, NewNetworkEvents, NewNetworkSender},
+    transport::ConnectionMetadata,
+    DisconnectReason, ProtocolId,
+};
+use rand::{rngs::StdRng, SeedableRng};
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    sync::Arc,
+};
+use storage_interface::mock::MockDbReader;
+use tokio::runtime::{Builder, Runtime};
+use vm_validator::mocks::mock_vm_validator::MockVMValidator;
+
+use crate::tests::node::{full_node, validator, vfn, Node, NodeTrait};
+use diem_config::config::NodeConfig;
+
+type NetworkHandle = (
+    NodeNetworkId,
+    MempoolNetworkSender,
+    NetworkEvents<MempoolSyncMsg>,
+);
+
+/// A test harness representing a combined network of Nodes and the mempool interactions between them
+#[derive(Default)]
+struct TestHarness {
+    mempools: HashMap<PeerId, Arc<Mutex<CoreMempool>>>,
+    network_reqs_rxs:
+        HashMap<PeerId, diem_channel::Receiver<(PeerId, ProtocolId), PeerManagerRequest>>,
+    network_notifs_txs:
+        HashMap<PeerId, diem_channel::Sender<(PeerId, ProtocolId), PeerManagerNotification>>,
+    network_conn_event_notifs_txs: HashMap<PeerId, conn_notifs_channel::Sender>,
+    runtimes: HashMap<PeerId, Runtime>,
+    subscribers: HashMap<PeerId, UnboundedReceiver<SharedMempoolNotification>>,
+    /// A mapping of secondary `PeerId` on other network interfaces to the primary `PeerId`
+    primary_peer_ids: HashMap<PeerId, PeerId>,
+}
+
+fn fifo_diem_channel<T, V>() -> (Sender<T, V>, Receiver<T, V>)
+where
+    T: Clone + Eq + Hash,
+{
+    static MAX_QUEUE_SIZE: usize = 8;
+    diem_channel::new(QueueStyle::FIFO, MAX_QUEUE_SIZE, None)
+}
+
+impl TestHarness {
+    /// Builds a validator only network for testing the SharedMempool interactions
+    fn bootstrap_validator_network(
+        validator_nodes_count: u32,
+        broadcast_batch_size: usize,
+        mempool_size: Option<usize>,
+        max_broadcasts_per_peer: Option<usize>,
+        max_ack_timeout: bool,
+    ) -> (TestHarness, Vec<Node>) {
+        let (harness, mut peers) = Self::bootstrap_network(
+            validator_nodes_count,
+            false,
+            0,
+            broadcast_batch_size,
+            mempool_size,
+            max_broadcasts_per_peer,
+            max_ack_timeout,
+        );
+        let validators = peers.remove(&PeerRole::Validator).unwrap();
+        (harness, validators)
+    }
+
+    /// Builds a fully functional network with Validators, attached VFNs, and full nodes
+    /// Note: None of these nodes are told about each other, and must manually be done afterwards
+    fn bootstrap_network(
```

Comment:
nit: I feel like this function could be split up a little to improve readability, i.e., we could have a helper for creating validators, validator fullnodes and another for fullnodes, as opposed to inlining all 3 segments of the code.

Reply:
Yeah, I agree it could be split up.  Originally, I had done a bunch of work to connect the validators to specific VFNs, and I ripped that out because it was too complicated.